src/test/ui/self/suggest-self.rs

```diff
@@ -0,0 +1,52 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+struct Foo {
+    x: i32,
+}
+
+impl Foo {
+    fn this1(&self) -> i32 {
+        let this = self;
+        let a = 1;
+        this.x
+    }
+
+    fn this2(&self) -> i32 {
+        let a = Foo {
+            x: 2
+        };
+        let this = a;
+        this.x
```

Comment:
@estebank the testcases `this1` `this2` passed, should it be marked as `Applicable`?

Reply:
@estebank I'm not sure if I get your points, here is my understanding of it and a maybeincorrect approach.
https://play.rust-lang.org/?gist=e73b1041b113b204ad930d1bff9035bb&version=nightly&mode=debug&edition=2015

Reply:
yep, added an entry to TODO.